### ISSUE TITLE
A suspicious misusage of realloc()?

### ISSUE THREAD

**Issue by movie-travel-code** [hercules-team/augeas#554]
Hi all, 
I'm henry wong, from Qihoo 360 Codesafe Team. We found **a suspicious misusage of realloc()** at https://github.com/hercules-team/augeas/blob/master/src/augtool.c#L162.

**`child[ctxidx]` may be accessing a released memory**. If `strlen(child)-strlen(ctx)+1` in `char *c = realloc(child, strlen(child)-strlen(ctx)+1);` is larger than memory length of `child`, the `realloc()` may allocate new memory, and release the `child`, at this time there will be a **accessing released memory**. Of course, this is probably not a problem. Since I'm not familiar with **augeas**, I think I should report this issue here to ask for your advice so as to improve **augeas**.

IMHO, the more appropriate approach is to put `child = c;` in front of the useage of `child`, as the code below.
```
    // ...
    char *c = realloc(child, strlen(child)-strlen(ctx)+1);
    if (c == NULL) {
        free(child);
            return NULL;
    }
    child = c;
    int ctxidx = strlen(ctx);
    if (child[ctxidx] == SEP)
        ctxidx++;
    strcpy(c, &child[ctxidx]);
    // child = c; 
    // ...
```

Thanks in advance!

Cheers!
Henry Wong
Qihoo 360 Codesafe Team

**Comment by lutter** [comment]
Hi,

thanks for the report - yes, you are absolutely right, there is a problem with the realloc; there's also a problem with using `strcpy` for possibly overlapping strings here. I need to think through the problems with this code a little more before coming up with a fix. I think there's also problems with how pointers that are stored in static variables are realloc'd but the static vars aren't updated.

**Comment by movie-travel-code** [comment]
@lutter sorry to bother, any progress?

**Comment by lutter** [comment]
Fixed by PR #574
